src/core/lib/security/transport/security_handshaker.c

```diff
@@ -321,18 +341,9 @@ static void on_handshake_data_sent_to_peer(grpc_exec_ctx *exec_ctx, void *arg,
     security_handshaker_unref(exec_ctx, h);
     return;
   }
-  /* We may be done. */
-  if (tsi_handshaker_is_in_progress(h->handshaker)) {
+  if (h->handshaker_result == NULL) {
```

Comment:
The only time we need to check_peer_locked() is inside on_handshake_next_done_locked(), when we get a non-empty handshaker_result. on_handshake_next_done_locked() is the only place security handshaker knows if the handshake completes. 

The if (h->handshaker_result == NULL) statement here is to decide whether we need to read from peer again.

In previous implementation, there are two places to check if handshake completes. When handshake completes, we do a check_peer_locked(). Thus, check_peer_locked() was called in two different places.